File: sdk/node/test/highwater.test.js
import assert from "node:assert/strict";
import test from "node:test";

import { HighWaterGuard, VerificationCode, isVerificationErrorCode } from "../src/index.js";

test("high-water guard tolerates small adjustment and rejects rollback", async () => {
  let saved = null;
  const store = {
    async load() {
      return saved;
    },
    async save(value) {
      saved = value;
    },
  };
  const guard = new HighWaterGuard({ store, allowedRollbackMs: 60_000 });
  const base = new Date("2026-08-12T03:00:00Z");
  await guard.checkAndUpdate(base);
  await guard.checkAndUpdate(new Date(base.getTime() - 30_000));
  await assert.rejects(
    guard.checkAndUpdate(new Date(base.getTime() - 120_000)),
    (error) => isVerificationErrorCode(error, VerificationCode.CLOCK_ROLLBACK),
  );
  const future = new Date(base.getTime() + 3_600_000);
  await guard.checkAndUpdate(future);
  assert.equal(saved.toISOString(), future.toISOString());
});

test("high-water guard serializes concurrent updates", async () => {
  let saved = null;
  const store = {
    async load() {
      await new Promise((resolve) => setTimeout(resolve, 2));
      return saved;
    },
    async save(value) {
      saved = value;
    },
  };
  const guard = new HighWaterGuard({ store });
  const first = new Date("2026-08-12T03:00:00Z");
  const second = new Date("2026-08-12T04:00:00Z");
  await Promise.all([guard.checkAndUpdate(first), guard.checkAndUpdate(second)]);
  assert.equal(saved.toISOString(), second.toISOString());
});

File: sdk/go/highwater_test.go
package ynlicense

import (
	"context"
	"testing"
	"time"
)

type memoryHighWaterStore struct {
	value time.Time
}

func (s *memoryHighWaterStore) Load(context.Context) (time.Time, error) {
	return s.value, nil
}

func (s *memoryHighWaterStore) Save(_ context.Context, value time.Time) error {
	s.value = value
	return nil
}

func TestHighWaterGuard(t *testing.T) {
	ctx := context.Background()
	store := &memoryHighWaterStore{}
	guard, err := NewHighWaterGuard(store, time.Minute)
	if err != nil {
		t.Fatalf("new guard: %v", err)
	}
	base := time.Date(2026, 8, 10, 3, 0, 0, 0, time.UTC)
	if err := guard.CheckAndUpdate(ctx, base); err != nil {
		t.Fatalf("first update: %v", err)
	}
	if err := guard.CheckAndUpdate(ctx, base.Add(-30*time.Second)); err != nil {
		t.Fatalf("small clock adjustment should be tolerated: %v", err)
	}
	if err := guard.CheckAndUpdate(ctx, base.Add(-2*time.Minute)); !IsVerificationErrorCode(err, VerificationClockRollback) {
		t.Fatalf("expected clock rollback, got %v", err)
	}
	if err := guard.CheckAndUpdate(ctx, base.Add(time.Hour)); err != nil {
		t.Fatalf("advance high water: %v", err)
	}
	if !store.value.Equal(base.Add(time.Hour)) {
		t.Fatalf("unexpected high-water value: %v", store.value)
	}
}
